# corgias data

The code to reproduce the results in Nishimura et al.
```
Yuki Nishimura, Kimiho Omae, Kento Tominnaga, Wataru Iwasaki.
CORGIAS: identifying correlated gene pairs by considering evolutionary history in a large-scale prokaryotic genome dataset
bioRxiv, 2025, https://doi.org/10.1101/2025.05.07.652372
```

## Requirments

- [CORGIAS](https://github.com/ynishimuraLv/corgias)
- [CheckM2](https://github.com/chklovski/CheckM2)
- [COGclassifier](https://github.com/moshi4/COGclassifier)
- [evolCCM](https://github.com/beiko-lab/evolCCM)
- [SynExtend(EvoWeaver)](https://bioconductor.org/packages/release/bioc/html/SynExtend.html)

## Files
**All `.py` files were generated by [Jupytext](https://github.com/mwouts/jupytext/tree/main), so can be converted to `.ipynb` and can run in [Jupyter](https://jupyter.org/)**

- These two files download genomes from [GTDB](https://gtdb.ecogenomic.org/) (Reelase 214), check their quality, annotate them, and performe ancestral state reconstruction and phylogenetic profiling.
    - 1_Prepare_dataset.py
        - run_checkM2.sh : A shell script file used in 1_Prepare_dataset.py
    - 2_phylogenetic_profiling.py
        - run_EvoWeaver.R: A R file used in 2_phylogenetic_profiling.py

**As these procedures needs a large comuptational resource, the resultant files required for downstream analysis can be downloaded from [Zenodo](10.5281/zenodo.15496141) (not available now)**

- Following files are for reproduction of Figures and Tables in Nishimura et al. `ipynb` files of them are available at [Zenodo](10.5281/zenodo.15496141)
    - 3_FigS1_TableS1.py
    - 4_Fig2_S2_TableS2-S7.py
    - 5_evolCCM_analysis.py
    - 6_Fig3_FigsS3-S5.py
    - 7_Fig4_FigS6-S10.py
    - 8_FigS11-S14.py
    - 9_Fig5_S15_S16_TableS8-S10.py
    - 10_Fig6_TableS11.py
        - Analysis of benchmark results using simulated data.
        - The code to simulate data (trees and profiles) and perform phylogenetic profiling can be found in ```simulation/```
        - simulation/
            - 1_simulate_pairs.py
            - 2_prepare.ipynb
            - 3_benchmark.ipynb
    - 11_Fig7_S17-S19.py
- [`cotransitions`](https://github.com/lab83bio/Cotransitions) reimplementation in CORGIAS is validated in a `Reimplement_cotr.ipynb`
    - A tree and table data in this file can be prepared `simulation/1_simulate_pairs.py`.



