# corgias data

The code to reproduce the results in Nishimura et al.
```
Yuki Nishimura, Kimiho Omae, Kento Tominnaga, Wataru Iwasaki.
CORGIAS: identifying correlated gene pairs by considering evolutionary history in a large-scale prokaryotic genome dataset
bioRxiv, 2025, https://doi.org/10.1101/2025.05.07.652372
```

## Requirments

- [CORGIAS](https://github.com/ynishimuraLv/corgias)
- [CheckM2](https://github.com/chklovski/CheckM2)
- [COGclassifier](https://github.com/moshi4/COGclassifier)
- [evolCCM](https://github.com/beiko-lab/evolCCM)

## Files
**All `.py` files were generated by [Jupytext](https://github.com/mwouts/jupytext/tree/main), so can be converted to `.ipynb` and can run in [Jupyter](https://jupyter.org/)**

- These two files download genomes from [GTDB](https://gtdb.ecogenomic.org/) (Reelase 214), check their quality, annotate them, and performe phylogenetic profiling.
    - 1_Prepare_dataset.py
        - run_checkM2.sh: A shell script file used in 1_Prepare_dataset.py
    - 2_phylogenetic_profiling.py

**As these procedures needs a large comuptational resource, the resultant files required for downstream analysis can be downloaded from [Zenodo](10.5281/zenodo.15496141)**

- Following files are for reproduction of Figures and Tables in Nishimura et al.
    - 3_TableS1.py
    - 4_Fig2_Table1_TableS3.py
    - 5_evolCCM_analysis.py
    - 6_Fig3_FigsS1-S5.py
    - 7_Fig4_FigS6_TablesS4-6.py
    - 8_Fig5_tableS7.py
        - Analysis of benchmark results using simulated data.
        - The code to simulate data (trees and profiles) and perform phylogenetic profiling can be found in ```simulation/```
    - simulation/
        - 1_simulate_pairs.py
        - 2_prepare.ipynb
        - 3_benchmark.ipynb


